### PR TITLE
Add a minimal API for ShadowMem clients

### PR DESCRIPTION
**Pull request by caballa**
This is the smallest API I can think of to use ShadowMem.
This API is enough for things like what Isabel is doing.

### CODE REVIEW

File: include/sea_dsa/ShadowMem.hh [hunk 1]
```diff
@@ -30,7 +49,24 @@ public:
 	    llvm::Pass &pass /* for dominatorTree and assumptionCache*/,
 	    bool splitDsaNodes = false, bool computeReadMod = false,
```
Comment (agurfinkel) [inline include/sea_dsa/ShadowMem.hh]: is `computeReadMod` used? It should be taken care of by `sea-dsa`. This was a limitation of `llvm-dsa`

File: include/sea_dsa/ShadowMem.hh [hunk 2]
```diff
@@ -30,7 +49,24 @@ public:
...
 	    bool memOptimizer = true, bool useTBAA = true);
```
Comment (agurfinkel) [inline include/sea_dsa/ShadowMem.hh]: TODO: `memOptimizer` should work in the presence of function calls. I think it does. There is just an assertion because that mode was untested.

File: include/sea_dsa/ShadowMem.hh [hunk 3]
```diff
@@ -30,7 +49,24 @@ public:
...
+
+  // Return the offset associated to cell c. The offset can be
+  // adjusted to zero if horn-sea-dsa-split=false.
+  unsigned getOffset(const Cell &c) const;
+
+  ShadowMemInstOp getShadowMemOp(const llvm::CallInst &ci) const; 
+  
+  // Return cell associated to the shadow mem call instruction.  It
+  // can be null if ci is not a shadow mem call instruction.
+  const Cell* getShadowMemCell(const llvm::CallInst &ci) const;
```
Comment (agurfinkel) [inline include/sea_dsa/ShadowMem.hh]: I don't remember if we pass `Cell` by value. If we do, this should be `llvm::Optional<Cell>`

File: include/sea_dsa/ShadowMem.hh [hunk 4]
```diff
@@ -30,7 +49,24 @@ public:
...
+
+  // Return a pair <def,use> with the variables which are defined
+  // and/or used by the shadow mem instruction. If the shadow
+  // instruction does not use or define a shadow variable then the
+  // corresponding pair element is null.
+  std::pair<llvm::Value*, llvm::Value*> getShadowMemVars(llvm::CallInst &ci) const;
```
Comment (agurfinkel) [inline include/sea_dsa/ShadowMem.hh]: I believe that there are up 3 variables. One def and 2 uses. This is in case of `mem.transfer`.

File: src/ShadowMem.cc [hunk 5]
```diff
@@ -253,14 +249,15 @@ class ShadowMemImpl : public InstVisitor<ShadowMemImpl> {
 
   // We use std::map to keep keys ordered.  
   using ShadowsMap =
-      std::map<const sea_dsa::Node *, std::map<unsigned, llvm::AllocaInst *>,
+      std::map<const dsa::Node *, std::map<unsigned, llvm::AllocaInst *>,
```
Comment (agurfinkel) [inline src/ShadowMem.cc]: we changed namespace from `sea_dsa` to `dsa`?

File: src/ShadowMem.cc [hunk 6]
```diff
@@ -275,14 +272,18 @@ class ShadowMemImpl : public InstVisitor<ShadowMemImpl> {
   /// \brief A map from DsaNode to its numeric id
   NodeIdMap m_nodeIds;
 
+  // Associate a shadow instruction to its Cell. Used by ShadowMem
+  // clients (e.g., VCGen).
+  llvm::DenseMap<const CallInst *, dsa::Cell> m_shadowMemInstToCell;
```
Comment (agurfinkel) [inline src/ShadowMem.cc]: Why is this necessary? A `dsa` cell can be easily retrieved from the `CallInst` by moving to the previous instruction.
This is also more robust, in case dsa analysis is re-run.

File: src/ShadowMem.cc [hunk 7]
```diff
@@ -450,6 +444,7 @@ class ShadowMemImpl : public InstVisitor<ShadowMemImpl> {
     ci = B.CreateCall(fn, {B.getInt32(getFieldId(c)), us}, "sm");
     markDefCall(ci, llvm::None);
     B.CreateStore(ci, a);
+    m_shadowMemInstToCell.insert({ci, dsa::Cell(c.getNode(), c.getOffset())});
```
Comment (agurfinkel) [inline src/ShadowMem.cc]: really dislike these calls everywhere

File: src/ShadowMem.cc [hunk 8]
```diff
@@ -652,10 +655,112 @@ class ShadowMemImpl : public InstVisitor<ShadowMemImpl> {
...
+  const Cell* getShadowMemCell(const CallInst &ci) const {
+    if (!isShadowMemInst(ci)) {
+      errs() << "ERROR: " << ci << " is not a shadow memory instruction.\n";
+      return nullptr;
+    }
+    auto it = m_shadowMemInstToCell.find(&ci);
+    if (it != m_shadowMemInstToCell.end()) {
+      return &(it->second);
+    } else {
+      errs() << "ERROR: cannot find cell associated to shadow mem inst "
```
Comment (agurfinkel) [inline src/ShadowMem.cc]: This is a warning. Should be silent by default.

File: src/ShadowMem.cc (hunk 5 above)
Comment (caballa) [reply src/ShadowMem.cc]: We have the alias `namespace dsa = sea_dsa;`. Sometimes we were using `dsa` and sometimes `sea_dsa`. I'm trying to be consistent and use always `dsa`

File: include/sea_dsa/ShadowMem.hh (hunk 4 above)
Comment (caballa) [reply include/sea_dsa/ShadowMem.hh]: Why two uses? I was considering the source pointer as an use and the destination as a def

File: src/ShadowMem.cc (hunk 5 above)
Comment (agurfinkel) [reply src/ShadowMem.cc]: ok

File: include/sea_dsa/ShadowMem.hh (hunk 4 above)
Comment (agurfinkel) [reply include/sea_dsa/ShadowMem.hh]: `memcpy(dst, src)` defines `dst` and uses both `dst` and `src`. `dst` is used to connect the previous def of `dst` to this one.

File: src/ShadowMem.cc (hunk 6 above)
Comment (caballa) [reply src/ShadowMem.cc]: I agree but not sure you can always do that. How can you get the cell from a`shadow.mem.init` or `shadow.mem.arg.ref` instruction ?

File: include/sea_dsa/ShadowMem.hh (hunk 1 above)
Comment (caballa) [reply include/sea_dsa/ShadowMem.hh]: If `sea-dsa` is context-insensitive then I think it still makes sense to use `computeReadMod`. But at least we should allow user to enable it only sea-dsa is context-insensitive

Comment (caballa) [reply include/sea_dsa/ShadowMem.hh]: Do you mean that the new flow-sensitivity of tea-dsa should subsume `computeReadMod`?

File: include/sea_dsa/ShadowMem.hh (hunk 4 above)
Comment (caballa) [reply include/sea_dsa/ShadowMem.hh]: `memcpy(dst, src)` is translated into two memory shadow instructions:
- `mem.shadow.trfr.load(src)`
- `sh = mem.shadow.store(dst)`
so in the end all memory shadow instruction have at most one definition and one use.
We can still change the API to allow more than one use per shadow mem instruction in the future, what do you think?
